keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,252 @@
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough, we may not need to request progress notify events from etcd.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
+should the kube-apiserver also wait 20ms to a desired revision to become
+available or should it fallback to making the request to etcd sooner?
+
+Optional: For some or all of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+### Risks and Mitigations
+
+This can increase the number of round trips to etcd: One is required to get the
+latest revision, another might be needed to request the progress notify. Both of
+these requests are cheap and scalable, but we need to understand the latency
+impact of waiting for both of them before serving a response back to the
+client. If this turns out to have performance implications, it could be
+partially mitigated by configuring etcd to send regular progress events, which
+is something we will explore.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
+request must be served from etcd anyway, which will be much higher latency than
+immediately serving from etcd like we do today. If this is an issue, it might be
+possible to partially mitigate it by tuning the “wait until fresh” duration to a
+sufficiently short value.
+
+## Design Details
+
+### Pagination
+
+Because the watch cache does not paginate responses, clients that were
+previously getting pagination for RV="" requests will start getting unpaginated
+responses when this feature is enabled.
```

Comment:
As I mentioned in [the comment](https://github.com/kubernetes/kubernetes/issues/86483#issuecomment-571404985), the most important point to consider is how to handle the page list

Reply:
I've added this to the graduation criteria so we don't go to beta until it is fully resolved.

I've also added a section about paging and have started to describe the problem. Let's continue to work on that and make sure we're satisfied with what we'd do for alpha before we mark this KEP implementable.